solver/src/main/java/org/chocosolver/solver/search/loop/monitors/NogoodFromRestarts.java

```diff
@@ -46,107 +52,110 @@
-                            png.addLearnt(Arrays.copyOf(lits, i + 1));
-                        }
-                    }
+            int lit = asLit(decision);
+            if (decision.hasNext() || decision.getArity() == 1) {
+                lits[i++] = lit;
+            } else {
+                if (i == 0) {
+                    // value can be removed permanently from var!
+                    png.addLearnt(lit);
                 } else {
-                    throw new UnsupportedOperationException("NogoodStoreFromRestarts cannot deal with such operator: " + ((SetDecision) decision).getDecOp());
+                    lits[i] = lit;
+                    png.addLearnt(Arrays.copyOf(lits, i + 1));
                 }
-            } else {
-                throw new UnsupportedOperationException("NogoodStoreFromRestarts can only deal with IntDecision and SetDecision.");
             }
         }
     }
+
+    /**
+     * Transform this decision into a literal to be used in {@link PropNogoods}.
+     *
+     * @param decision a decision
+     * @return the literal corresponding to this decision
+     */
+    private <V extends Variable> int asLit(Decision<V> decision) {
+        if (decision instanceof IntDecision) {
+            IntDecision id = (IntDecision) decision;
+            return asLit(
+                nogoodStealer.getById(id.getDecisionVariable(), png.getModel()), id.getDecOp(),
+                id.getDecisionValue()
+            );
+        } else if (decision instanceof SetDecision) {
+            SetDecision id = (SetDecision) decision;
+            return asLit(
+                nogoodStealer.getById(id.getDecisionVariable(), png.getModel()), id.getDecOp(),
+                id.getDecisionValue()
+            );
+        } else {
+            throw new UnsupportedOperationException("Cannot deal with such decision: " + decision);
+        }
+    }
+
+    private int asLit(IntVar var, DecisionOperator<IntVar> op, int val) {
+        int l;
+        if (DecisionOperatorFactory.makeIntEq().equals(op)) {
+            l = SatSolver.negated(png.Literal(var, val, true));
+        } else if (DecisionOperatorFactory.makeIntNeq().equals(op)) {
+            l = png.Literal(var, val, true);
+        } else if (DecisionOperatorFactory.makeIntSplit().equals(op)) {
+            l = SatSolver.negated(png.Literal(var, val, false));
+        } else if (DecisionOperatorFactory.makeIntReverseSplit().equals(op)) {
+            l = png.Literal(var, val, false);
+        } else {
+            throw new UnsupportedOperationException("Cannot deal with such operator: " + op);
+        }
+        return l;
+    }
+
+    private int asLit(SetVar var, DecisionOperator<SetVar> op, int val) {
+        int l;
+        if (op == DecisionOperatorFactory.makeSetForce()) {
+            l = SatSolver.negated(png.Literal(var, val, true));
+        } else if (op == DecisionOperatorFactory.makeSetRemove()) {
```

Comment:
Method `equals()` should be used instead of `==`.

Reply:
U're right